src/Plutarch/SafeMoney.hs

```diff
@@ -178,10 +187,99 @@ pdiscreteValue' (Tagged (AssetClass (cs, tn))) =
             PDiscrete t <- pmatchC p
             pure $ psingletonValue # pconstant cs # pconstant tn # (pextract # t)
 
-{-
-pdiscreteValue' (Tagged (AssetClass (cs, tn))) = phoistAcyclic $
-  plam $ \p ->
-    psingletonValue
-      # pconstant cs
-      # pconstant tn
-      # Tagged.puntag p -}
+-- | @since 1.0.1
+type PRateDecimal = PFixedDecimal 1000000
+
+-- | @since 1.0.1
+newtype PExchangeRate (from :: k) (to :: k') (s :: S)
+    = PExchangeRate (Term s (PTagged '(from, to) PRateDecimal))
+
+deriveGeneric ''PExchangeRate
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PlutusType (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PIsData (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PEq (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        POrd (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving anyclass instance PShow (PExchangeRate from to)
+
+-- | @since 1.0.1
+instance PTryFrom PData (PAsData (PExchangeRate from to)) where
+    type
+        PTryFromExcess PData (PAsData (PExchangeRate from to)) =
+            PTryFromExcess PData (PAsData PRateDecimal)
+    ptryFrom' d k = ptryFrom' @_ @(PAsData PRateDecimal) d $ k . first punsafeCoerce
+
+{- | Make @PExchangeRation@ with given @PIntegers. Two Integers are nominator and denominator.
+
+ @since 1.0.1
+-}
```

Comment:
It would be good to mention that we produce `PNothing` if the denominator is zero here.